### ISSUE TITLE
Failing test on local environment

### ISSUE THREAD

**Issue by stephenfin** [rst2pdf/rst2pdf#795]
I'm seeing a number of tests failing locally and I can't figure out why. Visually, the PDFs in `rst2pdf/tests/output` and `rst2pdf/tests/reference` look very similar.

- test_fancytitles.txt
- test_foreground.txt
- test_image_units_pdf.txt
- test_issue_263.txt
- test_issue_266_3.txt
- test_issue_266_4.txt
- test_issue_267.txt
- test_issue_294.txt
- test_issue_311.txt
- test_issue_383.txt
- test_vectorpdf.txt

The only common theme I see here is that the `vectorpdf` extension is used in many, though not all (I don't think `fancytitles` uses it) of them. I'm using my tox environment for this, so things are isolated, and I see the exact same failures with a virtualenv. I'm also running Fedora 29, but I have the exact same issues with an Ubuntu 16.04 VM. Any suggestions on where/how I can start investigating the issue?

```
$ pip freeze
aafigure==0.6
alabaster==0.7.12
attrs==19.1.0
Babel==2.7.0
backports.functools-lru-cache==1.5
certifi==2019.6.16
chardet==3.0.4
coverage==4.5.4
cssselect2==0.2.1
cycler==0.10.0
docutils==0.15.2
html5lib==1.0.1
idna==2.8
imagesize==1.1.0
Jinja2==2.10.1
kiwisolver==1.1.0
lxml==4.4.1
MarkupSafe==1.1.1
matplotlib==2.2.4
nose==1.3.7
numpy==1.16.4
packaging==19.1
pdfrw==0.4
Pillow==6.1.0
Pygments==2.4.2
pyparsing==2.4.2
PyPDF2==1.26.0
python-dateutil==2.8.0
pytz==2019.2
reportlab==3.5.23
requests==2.22.0
-e git+https://github.com/rst2pdf/rst2pdf@5223553be12428955bad5ee6b1aa09b353a672a1#egg=rst2pdf
six==1.12.0
smartypants==2.0.1
snowballstemmer==1.9.0
Sphinx==1.8.5
sphinxcontrib-websupport==1.1.2
subprocess32==3.5.4
svglib==0.9.2
tinycss2==0.6.1
typing==3.7.4
urllib3==1.25.3
webencodings==0.5.1
xhtml2pdf==0.2.3
```

**Comment by akrabat** [comment]
I suspect that it’s Inkscape related. It usually is when test_fancytitles is involved...

**Comment by stephenfin** [comment]
Any chance you'd be able to provide instructions to reproduce an environment in an Ubuntu VM, or something along those lines. If suspect it's a binary dependency too but I have nothing to diff against at the moment

**Comment by lornajane** [comment]
I am an Ubuntu user and I only have one failing test (fancy_titles!!) at the moment so I'll share my setup.

```
Ran 300 tests in 270.943s

FAILED (SKIP=26, failures=1)
```

Ubuntu 18.04, Python 3.7.0, Inkscape 0.92.3, here's `pip freeze`

```
aafigure==0.6
alabaster==0.7.12
Babel==2.6.0
backports.functools-lru-cache==1.5
certifi==2018.11.29
chardet==3.0.4
cycler==0.10.0
docutils==0.14
html5lib==1.0.1
idna==2.8
imagesize==1.1.0
Jinja2==2.10
kiwisolver==1.0.1
MarkupSafe==1.1.0
matplotlib==2.2.3
numpy==1.15.4
packaging==18.0
pdfrw==0.4
Pillow==5.3.0
Pygments==2.3.0
pyparsing==2.3.0
PyPDF2==1.26.0
python-dateutil==2.7.5
pytz==2018.7
reportlab==3.5.12
requests==2.21.0
six==1.12.0
smartypants==2.0.1
snowballstemmer==1.2.1
Sphinx==1.7.9
sphinxcontrib-websupport==1.1.0
subprocess32==3.5.3
svg2rlg==0.3
typing==3.6.6
urllib3==1.24.1
webencodings==0.5.1
xhtml2pdf==0.2.3
```

For the record, my fancytitles output looks visually fine and I get `Validity of file test_fancytitles.pdf checksum '651349f0143b26986707ef8a374a4d2b' is unknown.`. I'll try to run the compare script on it at some point.

Does some of that help? We could add `pip freeze` to the commands that run on Travis to get information about the setup there?

**Comment by stephenfin** [comment]
I created a virualenv manually, dumped that into a `requirements.txt` file, and ran `pip install -r requirements.txt` (followed by `pip install nose` and `pip install .` since it gave me an `OSError: [Errno 2] No such file or directory` warning otherwise), and got a whole bunch of _different_ failures :tada: :smile:

- test_background_svg.txt
- test_fancytitles.txt
- test_issue_244.txt
- test_issue_311.txt
- test_slides.txt
- test_svg.txt
- test_svg_alignment.txt
- sphinx-issue252

My `pip freeze` diff looks like below:

```
@@ -4,6 +4,7 @@
 backports.functools-lru-cache==1.5
 certifi==2018.11.29
 chardet==3.0.4
+coverage==4.5.4
 cycler==0.10.0
 docutils==0.14
 html5lib==1.0.1
@@ -13,6 +14,7 @@
 kiwisolver==1.0.1
 MarkupSafe==1.1.0
 matplotlib==2.2.3
+nose==1.3.7
 numpy==1.15.4
 packaging==18.0
 pdfrw==0.4
@@ -24,6 +26,8 @@
 pytz==2018.7
 reportlab==3.5.12
 requests==2.21.0
+# Editable install with no version control (rst2pdf==0.95.dev0)
+-e /tmp/test/lib/python3.7/site-packages
 six==1.12.0
 smartypants==2.0.1
 snowballstemmer==1.2.1
```

I'll try figure out the root cause of these issues and see what I can fix.

**Comment by stephenfin** [comment]
Okay, resolved a few of them by installing `svglib` (`pip install svglib`) along with the `texlive-ucs` binary dependencies (`sudo dnf install texlive-ucs`). I still see two failing tests:

- test_fancytitles.txt
- test_issue_311.txt
- sphinx-issue252

I'm still investigating the first of these but the second looks like it would never pass. If I diff the reference and output I see:

```
$ diff -u rst2pdf/tests/reference/test_issue_311.pdf rst2pdf/tests/output/test_issue_311.pdf                           
--- /home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/reference/test_issue_311.pdf    2019-07-16 11:23:07.474116339 +0100                                                                                
+++ /home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/output/test_issue_311.pdf       2019-08-19 17:19:59.053138092 +0100                                                                                
@@ -13,7 +13,7 @@                                                                                                                                                                                                  
 3 0 obj                                                                                                                                                                                                           
 <<                                                                                                                                                                                                                
 /A <<                                                                                                                                                                                                             
-/S /URI /Type /Action /URI (file:///Users/rob/Projects/python/rst2pdf/rst2pdf/tests/input/images/biohazard.pdf)                                                                                                   
+/S /URI /Type /Action /URI (file:///home/sfinucan/Development/rst2pdf/rst2pdf/tests/input/images/biohazard.pdf)                                                                                        
 >> /Border [ 0 0 0 ] /Rect [ 101.5929 753.0236 116.5929 765.0236 ] /Subtype /Link /Type /Annot                                                                                                                    
 >>                                                                                                                                                                                                                
 endobj                                                                                                                                                                                                            
```

(this is partial - there are more diffs but this is the first and most obvious one).

The third one is trickier. I see this:

```
Process "make" started on Mon Aug 19 17:42:47 2019                                                                                                                                                                 
...
  writing MyProject...                                                                                                                                                                                             
* [WARNING] basenodehandler.py:166 Unkn. node (self.textdispatch): sphinx.ext.mathbase.math [near line UNKNOWN in file UNKNOWN]                                                                                    
* [WARNING] basenodehandler.py:166 Unkn. node (self.elemdispatch): sphinx.ext.mathbase.displaymath [near line 13 in file /home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/input/sphinx-issue252/index.rs
t]                                                                                                                                                                                                                 
* [ERROR] pdfbuilder.py:144 format not resolved, probably missing URL scheme or undefined destination target for 'equation-euler'
* Traceback (most recent call last):
*   File "/home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/env/lib/python2.7/site-packages/rst2pdf/pdfbuilder.py", line 141, in write
*     docwriter.write(doctree, destination)
*   File "/home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/env/lib/python2.7/site-packages/docutils/writers/__init__.py", line 80, in write
*     self.translate()
*   File "/home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/env/lib/python2.7/site-packages/rst2pdf/pdfbuilder.py", line 635, in translate
*     compressed=self.compressed)
*   File "/home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/env/lib/python2.7/site-packages/rst2pdf/createpdf.py", line 641, in createPdf
*     pdfdoc.multiBuild(elements)
*   File "/home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/env/lib/python2.7/site-packages/reportlab/platypus/doctemplate.py", line 1133, in multiBuild
*     self.canv.save()
*   File "/home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/env/lib/python2.7/site-packages/reportlab/pdfgen/canvas.py", line 1240, in save
*     self._doc.SaveToFile(self._filename, self)
*   File "/home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/env/lib/python2.7/site-packages/reportlab/pdfbase/pdfdoc.py", line 215, in SaveToFile
*     data = self.GetPDFData(canvas)
*   File "/home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/env/lib/python2.7/site-packages/reportlab/pdfbase/pdfdoc.py", line 241, in GetPDFData
*     return self.format()
*   File "/home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/env/lib/python2.7/site-packages/reportlab/pdfbase/pdfdoc.py", line 418, in format
*     IOf = IO.format(self)
*   File "/home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/env/lib/python2.7/site-packages/reportlab/pdfbase/pdfdoc.py", line 866, in format
*     fcontent = format(self.content, document, toplevel=1)   # yes this is at top level
*   File "/home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/env/lib/python2.7/site-packages/reportlab/pdfbase/pdfdoc.py", line 80, in format
*     f = element.format(document)
*   File "/home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/env/lib/python2.7/site-packages/reportlab/pdfbase/pdfdoc.py", line 1602, in format
*     return D.format(document)
*   File "/home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/env/lib/python2.7/site-packages/reportlab/pdfbase/pdfdoc.py", line 674, in format
*     L = [(format(PDFName(k),document)+b" "+format(dict[k],document)) for k in keys]
*   File "/home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/env/lib/python2.7/site-packages/reportlab/pdfbase/pdfdoc.py", line 80, in format
*     f = element.format(document)
*   File "/home/sfinucan/Development/innovation/rst2pdf/rst2pdf/tests/env/lib/python2.7/site-packages/reportlab/pdfbase/pdfdoc.py", line 1767, in format
*     if f is None: raise ValueError("format not resolved, probably missing URL scheme or undefined destination target for '%s'" % self.name)
* ValueError: format not resolved, probably missing URL scheme or undefined destination target for 'equation-euler'
  FAILED
  build succeeded.
   
  Build finished. The PDF files are in _build/pdf.
```

I have no idea how to resolve this :disappointed:

**Comment by stephenfin** [comment]
At least on my Fedora 31 host, most of these issues have been resolved by the move to content-based comparison in #854 and the remaining failures are addressed by #869. We can close this now :raised_hands:
